### ISSUE TITLE
GitLab Registry integration

### ISSUE THREAD

**Issue by rei-ifesca** [debops/debops#556]
Hi,

we have manage to configure and run a Docker Registry within our GitLab server. This registry service can be accessed over port 5000 and 443, thanks to simple nginx virtual host proxy configurations ;)

Due to lack of good documentation on the Docker and GitLab side it was not easy to make each other talk... Nevertheless, we would like to return our knowledge to the community and extend the debops.gitlab role to support a docker registry.

Because of the circular depenencies between the docker registry container and gitlab and additonal dependencies to the gitlab role (e.g. debops.docker) I would like to discuss the best path for a nice PR ;)

I have the following ideas:
- Extend the gitlab role documentation and templates to configure the registry, including PKI rootbundle support
- Add a debops.dsocker_registry role to deploy a docker-registry container over compose with ready-to-go configuration to work together with gitlab, however also support a standalone docker registry service over TLS with simple HTTP auth.

Any opinions?

**Comment by drybjed** [comment]
Definitely the direction I'd like to see this go in. At the moment I have a working Docker Registry role on a separate branch in https://github.com/drybjed/debops/tree/wip-docker_registry/ansible/roles/debops.docker_registry with a corresponding playbook at https://github.com/drybjed/debops/blob/wip-docker_registry/ansible/playbooks/service/docker_registry.yml. It compiles and installs Docker Registry locally, without the need for Docker itself, and uses `debops.nginx` as a frontend. I think what it lacks currently is documentation (obviously) and a way to clean up the registry periodically to avoid disk space issues; I think that should be done wia a `cron` job, not sure yet what exactly. The configuration scheme seems reasonable, users should be able to configure anything they need via Ansible inventory.

I'm not really fond of using the Docker Registry in a container on top of Docker itself, that seems as an unneeded dependency on Docker in this case. Additional bonus is having Docker Registry source code available locally, in case it ever went away. The role could also be updated later to use [the `docker-registry` packaged in Debian Buster](https://packages.debian.org/buster/docker-registry) when it becomes Stable, as an alternative.

Did you manage to setup the Registry with GitLab on separate hosts, or do they have to live together on the same host? In the second case, utilizing Ansible local facts to pass configuration between the roles might come in handy, in the first case inventory will have to be utilized to faciliate coordination between the hosts - this could be further automated later with something like Consul, but I don't want to introduce dependencies to the project that are not manageable by DebOps.

**Comment by rei-ifesca** [comment]
> Definitely the direction I'd like to see this go in. At the moment I have a working Docker Registry role on a separate branch  

:+1: 

> It compiles and installs Docker Registry locally, without the need for Docker itself, and uses `debops.nginx` as a frontend. 

This is a good solution where docker is an unwanted dependency.

> I think that should be done wia a `cron` job, not sure yet what exactly. 

Maybe several wiping configurations depending on age of the image and tag name?

> I'm not really fond of using the Docker Registry in a container on top of Docker itself, that seems as an unneeded dependency on Docker in this case.

In some use case a docker environment is given and used, so I would allow the admin to choose between a source_code or a docker_compose installation. 
Beside the needed docker daemon is not much to change in the role.

> The role could also be updated later to use [the `docker-registry` packaged in Debian Buster](https://packages.debian.org/buster/docker-registry) when it becomes Stable, as an alternative.

Good hint, this makes of course everything easier.

> Did you manage to setup the Registry with GitLab on separate hosts, or do they have to live together on the same host? 

The registry redis currently in the same host as gitlab. The official GitLab documentation is ambiguous if is a good idea to split the services into several hosts. Technically possible, however the config variable "registry.path" tells something else.
Currently our use case works with both services in the same host machine, thus no need for us to change, yet.

> In the second case, utilizing Ansible local facts to pass configuration between the roles might come in handy, in the first case inventory will have to be utilized to faciliate coordination between the hosts -

Local facts: I forgot about this central wisdom base ;) Ideal for this use case with circular dependencies.

>  this could be further automated later with something like Consul, but I don't want to introduce dependencies to the project that are not manageable by DebOps.

One step at a time ;)

1. docker registry configurable as independent role either with or without docker
2. connect the gitlab role to the docker registry. Config stuff. Almost done with that part ;)
3. Extend the registry role

**Comment by drybjed** [comment]
> Maybe several wiping configurations depending on age of the image and tag name?

Sure, sounds good. It could even be configurable by the role.

> > I'm not really fond of using the Docker Registry in a container on top of Docker itself, that seems as an unneeded dependency on Docker in this case.
> 
> In some use case a docker environment is given and used, so I would allow the admin to choose between a source_code or a docker_compose installation.
> Beside the needed docker daemon is not much to change in the role.

Sure. I guess the `debops.docker_registry` role could check if Docker is present via `debops.docker` Ansible local facts. But, the issue might be using the `debops.docker` setup via Ansible, because [this requires custom Python interpreter inside of the `virtualenv` set up by `debops.docker`](https://docs.debops.org/en/master/ansible/roles/debops.docker/docker-virtualenv.html#ansible-modules-and-docker-virtualenv). In essence it would have to be done via a separate playbook. I guess the alternative, `docker-compose`-based approach could be done as a "sub-role" similar to how some DebOps roles handle the environment with `/env` subdirectory. Then the same role can be included in two playbooks, working a bit differently, but using the same defaults.

> > Did you manage to setup the Registry with GitLab on separate hosts, or do they have to live together on the same host?
> 
> The registry redis currently in the same host as gitlab. The official GitLab documentation is ambiguous if is a good idea to split the services into several hosts. Technically possible, however the config variable "registry.path" tells something else.
> Currently our use case works with both services in the same host machine, thus no need for us to change, yet.

I guess it can't be helped for now. However, I wouldn't want to include `debops.docker_registry` as a dependency in `debops.gitlab` playbook, let's keep them separate and use facts for communication. This will require multiple passes of the `debops.gitlab` playbook to complete the setup (ie. GitLab, Registry, GitLab again), but I would be fine with that for the moment.

I suppose that a NFS share could solve the issue with multi-host deployments.

> > In the second case, utilizing Ansible local facts to pass configuration between the roles might come in handy, in the first case inventory will have to be utilized to faciliate coordination between the hosts -
> 
> Local facts: I forgot about this central wisdom base ;) Ideal for this use case with circular dependencies.

Yes, they are handy. At the moment, `debops.docker_registry` doesn't have facts, do you know what information would have to be exposed for `debops.gitlab` to configure everything that's needed? I imagine that the passphrase would have to be shared via the `secret/` directory as well, so that it can be randomly generated and passed between the hosts later one, with multi-host deployment.

> > this could be further automated later with something like Consul, but I don't want to introduce dependencies to the project that are not manageable by DebOps.
> 
> One step at a time ;)
> 1. docker registry configurable as independent role either with or without docker
> 2. connect the gitlab role to the docker registry. Config stuff. Almost done with that part ;)
> 3. Extend the registry role

Agreed. So, do you want to get the existing `debops.docker_registry` role and clean it up, or should I finish documenting it, merge to DebOps `master` and then we can work on GitLab integration?

**Comment by rei-ifesca** [comment]
> I guess the alternative, `docker-compose`-based approach could be done as a "sub-role" similar to how some DebOps roles handle the environment with `/env` subdirectory. Then the same role can be included in two playbooks, working a bit differently, but using the same defaults.

I would rather prefer not to split the role that way, otherwise we need to maintain two file trees. 
A debops.docker_registry role needs the following prior-knowledge in order to work:

- enable docker registry
- which directories use to store the data
- which authentication method?
- which TCP ports to use
- use nginx as (TLS) proxy?
  - if yes, make standard pki (acme) stuff
- use nginx as HTTP Access method?
- use docker daemon 
  - if yes:
    - is the docker daemon available and running?
    - using docker-compose either over the special docker-python virtualenv or directly deploy the registry. See simple working example attached.
  - no:
    -  install registry as binary or source code or .deb 


> However, I wouldn't want to include `debops.docker_registry` as a dependency in `debops.gitlab` playbook, let's keep them separate and use facts for communication.

Definitely.

> This will require multiple passes of the `debops.gitlab` playbook to complete the setup (ie. GitLab, Registry, GitLab again), but I would be fine with that for the moment.

I don't think this is necessary. By managing logic in the roles it should be possible to fail for example when docker or the registry are not installed, but selected by the user.
The services playbook need to take into account, that docker and docker_registry must run before gitlab.

> I suppose that a NFS share could solve the issue with multi-host deployments.

With proper documentation explaining why, an admin should be able to make the decision on how to manage this common path (e.g. NFS, CEPH, S3, sshfs, ...) I would prefer not add any logic to the role here. Just a path to the filesystem, if configured correctly should not matter for gitlab and/or the registry., 
We use for this scenario a mounted network volume from our cloud provider, works like a charm. 
Although mounted mit debops.nfs (just to modify the /etc/fstab ^^) and provisioned with debops.resources.


> Yes, they are handy. At the moment, `debops.docker_registry` doesn't have facts, do you know what information would have to be exposed for `debops.gitlab` to configure everything that's needed? 

I will push my branch with alpha changes to my debops fork containing the config and parameters are that needed. I will also provide an Example docker-registry playbook.

> I imagine that the passphrase would have to be shared via the `secret/` directory as well, so that it can be randomly generated and passed between the hosts later one, with multi-host deployment.

ACK

> Agreed. So, do you want to get the existing `debops.docker_registry` role and clean it up, or should I finish documenting it, merge to DebOps `master` and then we can work on GitLab integration?

You know your role better, I think it is better that you improve the documentation on the role. I could then provide PRs to improve it before merging.

=== Example===

Working example of a docker registry ready configured to run together with GitLab including user authentication over GitLab. In Addition the gitlab config is provided.
```
---

- name: Install Docker registry
  hosts: [ 'gitlab.example.com' ]
  tags: [ 'play::management:tools:docker-registry' ]
  become: yes

  # See: https://docs.gitlab.com/ee/administration/container_registry.html
  # ToDo: alter config file /var/local/git/gitlab/config/gitlab.yml
  #  registry:
  #    enabled: true
  #    host: "registry.example.com"
  #    #port: 443
  #    api_url: "http://localhost:5005/"
  #    key: "/etc/pki/realms/gitlab.example.com/default.key"
  #    path: "/mnt/docker/registry"
  #    issuer: "gitlab-issuer"

  vars:
    # https://docs.debops.org/en/master/ansible/roles/debops.docker/docker-virtualenv.html
    # Use Python from Docker virtualenv
    ansible_python_interpreter: '/usr/bin/env docker-python'

    docker_registry_listen_port: 5005
    docker_registry_host_data_path: '/mnt/docker/registry'
    docker_registry_rootcertbundle: /etc/pki/realms/gitlab.example.com/default.crt'

  tasks:

  # See: https://docs.ansible.com/ansible/latest/modules/docker_service_module.html
  # See: https://docs.docker.com/registry/deploying/#deploy-your-registry-using-a-compose-file
  # See: https://docs.docker.com/registry/configuration/#auth
  - name: Start Docker Registry
    docker_service:
      project_name: registry.example.com
      definition:
        version: '2'
        services:
          registry:
            restart: always
            image: registry:2
            ports:
              - '{{ docker_registry_listen_port }}:5000'
            environment:
              REGISTRY_STORAGE_FILESYSTEM_ROOTDIRECTORY: '/var/lib/registry'
              REGISTRY_AUTH: 'token'
              REGISTRY_AUTH_TOKEN_REALM: 'https://gitlab.example.com/jwt/auth'
              REGISTRY_AUTH_TOKEN_SERVICE: 'container_registry'
              REGISTRY_AUTH_TOKEN_ISSUER: 'gitlab-issuer'
              REGISTRY_AUTH_TOKEN_ROOTCERTBUNDLE: '/certs/default.crt'
              REGISTRY_STORAGE_DELETE_ENABLED: 'true'
            volumes:
              - '{{ docker_registry_host_data_path }}:/var/lib/registry'
              - '{{ docker_registry_rootcertbundle }}:/certs:ro'

...
```

**Comment by rei-ifesca** [comment]
Just in case you wondered, yes the "registry.port" in the config is commented, it is optional and just cosmetics to tell the docker registry user which port to use. In our case we use port 443, thus as standard HTTPS port does not need to be specified ;)

**Comment by drybjed** [comment]
> > I guess the alternative, `docker-compose`-based approach could be done as a "sub-role" similar to how some DebOps roles handle the environment with `/env` subdirectory. Then the same role can be included in two playbooks, working a bit differently, but using the same defaults.
> 
> I would rather prefer not to split the role that way, otherwise we need to maintain two file trees.

It's not that bad. The sub-role will have just one `docker_service` task, basically something like the one you posted, but with various parameters exposed via defaults so that they can be modified via inventory variables. The sub-role `defaults/main.yml` will be a symlink to the main role's `defaults/main.yml`.

The sub-role will be executed in its own play, with its own separate inventory group (perhaps something like `[debops_service_docker_registry_in_docker]` and custom Ansible interpreter pointing to the Docker `virtualenv`. I actually went with this approach initially but then decided that I liked a locally installed Registry better. I'll do both in the role.

> > This will require multiple passes of the `debops.gitlab` playbook to complete the setup (ie. GitLab, Registry, GitLab again), but I would be fine with that for the moment.
> 
> I don't think this is necessary. By managing logic in the roles it should be possible to fail for example when docker or the registry are not installed, but selected by the user.
> The services playbook need to take into account, that docker and docker_registry must run before gitlab.

I see. I guess that order should be fine then.

> > I suppose that a NFS share could solve the issue with multi-host deployments.
> 
> With proper documentation explaining why, an admin should be able to make the decision on how to manage this common path (e.g. NFS, CEPH, S3, sshfs, ...) I would prefer not add any logic to the role here. Just a path to the filesystem, if configured correctly should not matter for gitlab and/or the registry.,
> We use for this scenario a mounted network volume from our cloud provider, works like a charm.
> Although mounted mit debops.nfs (just to modify the /etc/fstab ^^) and provisioned with debops.resources.

Of course, separation of concerns is essential. `debops.docker_registry` does not have to know anything about the filesystem to work.

> > Yes, they are handy. At the moment, `debops.docker_registry` doesn't have facts, do you know what information would have to be exposed for `debops.gitlab` to configure everything that's needed?
> 
> I will push my branch with alpha changes to my debops fork containing the config and parameters are that needed. I will also provide an Example docker-registry playbook.

Thanks, that will be useful.

> > Agreed. So, do you want to get the existing `debops.docker_registry` role and clean it up, or should I finish documenting it, merge to DebOps `master` and then we can work on GitLab integration?
> 
> You know your role better, I think it is better that you improve the documentation on the role. I could then provide PRs to improve it before merging.

Great. I'll try to finish the role within a week.

**Comment by drybjed** [comment]
> `tags: [ 'play::management:tools:docker-registry' ]`

Interesting tag. Care to expand a bit about the scheme, perhaps with some more examples?

**Comment by rei-ifesca** [comment]
> Interesting tag. Care to expand a bit about the scheme, perhaps with some more examples?

It is actually based on the tag scheme from debops :-D
Like a reverse DNS:

- `play`: it's a playbook
- `management`: part of the management infrastructure
- `tools`: same idea, its a tool
- `docker-registry`: self explanatory

I hope it helps

**Comment by rei-ifesca** [comment]
Here are my gitlab role changes: https://github.com/rei-ifesca/debops/commit/534fa2d5f9e595b8afe893a0a45a4bd5f1fcd4f6

**Comment by drybjed** [comment]
@rei-ifesca The Docker Registry support is now getting ready to be merged in PR #667. I'll merge it most likely tomorrow, after testing an install from scratch.

**Comment by rei-ifesca** [comment]
nice!
